File: app/src/main/java/com/runtracker/app/ui/screens/gamification/GamificationScreen.kt
package com.runtracker.app.ui.screens.gamification

import androidx.compose.animation.core.*
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.runtracker.shared.data.model.*
import java.text.SimpleDateFormat
import java.util.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun GamificationScreen(
    viewModel: GamificationViewModel = hiltViewModel(),
    onNavigateToWellness: () -> Unit = {},
    onBack: () -> Unit
) {
    val uiState by viewModel.uiState.collectAsState()
    val newAchievements by viewModel.newAchievements.collectAsState()
    
    // Show achievement unlock dialog
    newAchievements.firstOrNull()?.let { achievement ->
        AchievementUnlockedDialog(
            achievement = achievement,
            onDismiss = { viewModel.dismissAchievementNotification(achievement.id) }
        )
    }
    
    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Progress & Achievements") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Back")
                    }
                }
            )
        }
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding),
            contentPadding = PaddingValues(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            // Level & XP Card
            item {
                uiState.gamification?.let { gamification ->
                    LevelCard(gamification = gamification)
                }
            }
            
            // Daily Rings
            item {
                uiState.todayRings?.let { rings ->
                    DailyRingsCard(rings = rings)
                }
            }
            
            // Streak Card
            item {
                uiState.gamification?.let { gamification ->
                    StreakCard(gamification = gamification)
                }
            }
            
            // Weekly Progress
            item {
                WeeklyProgressCard(weeklyRings = uiState.weeklyRings)
            }
            
            // Achievements Section
            item {
                Text(
                    "Achievements",
                    style = MaterialTheme.typography.titleLarge,
                    fontWeight = FontWeight.Bold
                )
            }
            
            // Achievement categories
            uiState.achievementsByCategory.forEach { (category, achievements) ->
                item {
                    AchievementCategorySection(
                        category = category,
                        achievements = achievements
                    )
                }
            }
            
            // Recent XP
            if (uiState.recentXp.isNotEmpty()) {
                item {
                    RecentXpCard(transactions = uiState.recentXp)
                }
            }
        }
    }
}

@Composable
fun LevelCard(gamification: UserGamification) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.primaryContainer
        )
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(20.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            // Level badge
            Box(
                modifier = Modifier
                    .size(80.dp)
                    .clip(CircleShape)
                    .background(MaterialTheme.colorScheme.primary),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "${gamification.currentLevel}",
                    style = MaterialTheme.typography.headlineLarge,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onPrimary
                )
            }
            
            Spacer(modifier = Modifier.height(8.dp))
            
            Text(
                text = "Level ${gamification.currentLevel}",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold
            )
            
            Text(
                text = "${gamification.totalXp} XP",
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onPrimaryContainer.copy(alpha = 0.8f)
            )
            
            Spacer(modifier = Modifier.height(12.dp))
            
            // Progress to next level
            Column(modifier = Modifier.fillMaxWidth()) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Text(
                        text = "Next Level",
                        style = MaterialTheme.typography.labelSmall
                    )
                    Text(
                        text = "${gamification.xpToNextLevel} XP to go",
                        style = MaterialTheme.typography.labelSmall
                    )
                }
                Spacer(modifier = Modifier.height(4.dp))
                LinearProgressIndicator(
                    progress = gamification.levelProgress,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(8.dp)
                        .clip(RoundedCornerShape(4.dp)),
                    color = MaterialTheme.colorScheme.primary,
                    trackColor = MaterialTheme.colorScheme.onPrimaryContainer.copy(alpha = 0.2f)
                )
            }
        }
    }
}

@Composable
fun DailyRingsCard(rings: DailyRings) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "Today's Goals",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold
            )
            
            Spacer(modifier = Modifier.height(16.dp))
            
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceEvenly,
                verticalAlignment = Alignment.CenterVertically
            ) {
                // Activity Ring
                ActivityRing(
                    progress = rings.moveProgress,
                    color = Color(0xFFFF2D55),
                    label = "Move",
                    value = "${rings.moveMinutes}/${rings.moveGoal}",
                    unit = "min"
                )
                
                // Exercise Ring
                ActivityRing(
                    progress = rings.exerciseProgress,
                    color = Color(0xFF30D158),
                    label = "Exercise",
                    value = "${rings.exerciseMinutes}/${rings.exerciseGoal}",
                    unit = "min"
                )
                
                // Distance Ring
                ActivityRing(
                    progress = rings.distanceProgress,
                    color = Color(0xFF007AFF),
                    label = "Distance",
                    value = "%.1f/%.1f".format(rings.distanceMeters / 1000, rings.distanceGoal / 1000),
                    unit = "km"
                )
            }
        }
    }
}

@Composable
fun ActivityRing(
    progress: Float,
    color: Color,
    label: String,
    value: String,
    unit: String,
    modifier: Modifier = Modifier
) {
    val animatedProgress by animateFloatAsState(
        targetValue = progress,
        animationSpec = tween(1000, easing = FastOutSlowInEasing),
        label = "ring_progress"
    )
    
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = modifier
    ) {
        Box(
            modifier = Modifier.size(80.dp),
            contentAlignment = Alignment.Center
        ) {
            Canvas(modifier = Modifier.fillMaxSize()) {
                val strokeWidth = 10.dp.toPx()
                val radius = (size.minDimension - strokeWidth) / 2
                val center = Offset(size.width / 2, size.height / 2)
                
                // Background ring
                drawCircle(
                    color = color.copy(alpha = 0.2f),
                    radius = radius,
                    center = center,
                    style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
                )
                
                // Progress ring
                drawArc(
                    color = color,
                    startAngle = -90f,
                    sweepAngle = 360f * animatedProgress.coerceAtMost(1f),
                    useCenter = false,
                    topLeft = Offset(center.x - radius, center.y - radius),
                    size = Size(radius * 2, radius * 2),
                    style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
                )
            }
            
            // Percentage or checkmark
            if (progress >= 1f) {
                Icon(
                    Icons.Default.Check,
                    contentDescription = "Complete",
                    tint = color,
                    modifier = Modifier.size(24.dp)
                )
            } else {
                Text(
                    text = "${(progress * 100).toInt()}%",
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.Bold,
                    color = color
                )
            }
        }
        
        Spacer(modifier = Modifier.height(4.dp))
        
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            fontWeight = FontWeight.Medium
        )
        Text(
            text = value,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
fun StreakCard(gamification: UserGamification) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = if (gamification.currentStreak > 0) 
                Color(0xFFFF9500).copy(alpha = 0.15f) 
            else 
                MaterialTheme.colorScheme.surfaceVariant
        )
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            // Fire icon
            Box(
                modifier = Modifier
                    .size(56.dp)
                    .clip(CircleShape)
                    .background(Color(0xFFFF9500).copy(alpha = 0.2f)),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "🔥",
                    fontSize = 28.sp
                )
            }
            
            Spacer(modifier = Modifier.width(16.dp))
            
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = "${gamification.currentStreak} Day Streak",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold
                )
                Text(
                    text = if (gamification.currentStreak > 0) 
                        "Keep it going!" 
                    else 
                        "Start your streak today!",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            
            Column(horizontalAlignment = Alignment.End) {
                Text(
                    text = "Best",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text(
                    text = "${gamification.longestStreak}",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}

@Composable
fun WeeklyProgressCard(weeklyRings: List<DailyRings>) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "This Week",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold
            )
            
            Spacer(modifier = Modifier.height(12.dp))
            
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                val dayNames = listOf("M", "T", "W", "T", "F", "S", "S")
                val calendar = Calendar.getInstance()
                calendar.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY)
                
                dayNames.forEachIndexed { index, dayName ->
                    val dayStart = calendar.timeInMillis
                    val ringForDay = weeklyRings.find { 
                        val ringCal = Calendar.getInstance()
                        ringCal.timeInMillis = it.date
                        ringCal.get(Calendar.DAY_OF_YEAR) == calendar.get(Calendar.DAY_OF_YEAR)
                    }
                    
                    DayIndicator(
                        dayName = dayName,
                        isCompleted = ringForDay?.allRingsClosed == true,
                        hasActivity = ringForDay != null && ringForDay.exerciseMinutes > 0,
                        isToday = Calendar.getInstance().get(Calendar.DAY_OF_WEEK) == 
                                  (if (index == 6) Calendar.SUNDAY else Calendar.MONDAY + index)
                    )
                    
                    calendar.add(Calendar.DAY_OF_MONTH, 1)
                }
            }
        }
    }
}

@Composable
fun DayIndicator(
    dayName: String,
    isCompleted: Boolean,
    hasActivity: Boolean,
    isToday: Boolean
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            text = dayName,
            style = MaterialTheme.typography.labelSmall,
            color = if (isToday) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant
        )
        Spacer(modifier = Modifier.height(4.dp))
        Box(
            modifier = Modifier
                .size(32.dp)
                .clip(CircleShape)
                .background(
                    when {
                        isCompleted -> Color(0xFF30D158)
                        hasActivity -> Color(0xFFFF9500).copy(alpha = 0.5f)
                        isToday -> MaterialTheme.colorScheme.primaryContainer
                        else -> MaterialTheme.colorScheme.surfaceVariant
                    }
                ),
            contentAlignment = Alignment.Center
        ) {
            if (isCompleted) {
                Icon(
                    Icons.Default.Check,
                    contentDescription = "Completed",
                    tint = Color.White,
                    modifier = Modifier.size(16.dp)
                )
            }
        }
    }
}

@Composable
fun AchievementCategorySection(
    category: AchievementCategory,
    achievements: List<AchievementWithProgress>
) {
    val categoryName = category.name.lowercase().replaceFirstChar { it.uppercase() }
    val unlockedCount = achievements.count { it.isUnlocked }
    
    Column {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = categoryName,
                style = MaterialTheme.typography.titleSmall,
                fontWeight = FontWeight.Bold
            )
            Text(
                text = "$unlockedCount/${achievements.size}",
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
        
        Spacer(modifier = Modifier.height(8.dp))
        
        LazyRow(
            horizontalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            items(achievements.sortedByDescending { it.isUnlocked }) { achievementWithProgress ->
                AchievementBadge(achievementWithProgress = achievementWithProgress)
            }
        }
    }
}

@Composable
fun AchievementBadge(achievementWithProgress: AchievementWithProgress) {
    val achievement = achievementWithProgress.achievement
    val isUnlocked = achievementWithProgress.isUnlocked
    
    Card(
        modifier = Modifier.width(100.dp),
        colors = CardDefaults.cardColors(
            containerColor = if (isUnlocked) 
                MaterialTheme.colorScheme.primaryContainer 
            else 
                MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f)
        )
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(12.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            // Badge icon
            Box(
                modifier = Modifier
                    .size(48.dp)
                    .clip(CircleShape)
                    .background(
                        if (isUnlocked) 
                            MaterialTheme.colorScheme.primary 
                        else 
                            MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.3f)
                    ),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    when (achievement.category) {
                        AchievementCategory.DISTANCE -> Icons.Default.DirectionsRun
                        AchievementCategory.WORKOUTS -> Icons.Default.FitnessCenter
                        AchievementCategory.STREAKS -> Icons.Default.LocalFireDepartment
                        AchievementCategory.SPEED -> Icons.Default.Speed
                        AchievementCategory.CONSISTENCY -> Icons.Default.DateRange
                        AchievementCategory.PERSONAL_BEST -> Icons.Default.EmojiEvents
                        else -> Icons.Default.Star
                    },
                    contentDescription = null,
                    tint = if (isUnlocked) MaterialTheme.colorScheme.onPrimary else Color.Gray,
                    modifier = Modifier.size(24.dp)
                )
            }
            
            Spacer(modifier = Modifier.height(8.dp))
            
            Text(
                text = achievement.name,
                style = MaterialTheme.typography.labelSmall,
                fontWeight = FontWeight.Medium,
                textAlign = TextAlign.Center,
                maxLines = 2,
                color = if (isUnlocked) 
                    MaterialTheme.colorScheme.onPrimaryContainer 
                else 
                    MaterialTheme.colorScheme.onSurfaceVariant
            )
            
            if (!isUnlocked) {
                Spacer(modifier = Modifier.height(4.dp))
                LinearProgressIndicator(
                    progress = achievementWithProgress.progressPercent,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(4.dp)
                        .clip(RoundedCornerShape(2.dp)),
                )
            }
        }
    }
}

@Composable
fun RecentXpCard(transactions: List<XpTransaction>) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = "Recent XP",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold
            )
            
            Spacer(modifier = Modifier.height(8.dp))
            
            transactions.take(5).forEach { transaction ->
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 4.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = transaction.reason.name.replace("_", " ").lowercase()
                            .replaceFirstChar { it.uppercase() },
                        style = MaterialTheme.typography.bodySmall
                    )
                    Text(
                        text = "+${transaction.amount} XP",
                        style = MaterialTheme.typography.bodySmall,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.primary
                    )
                }
            }
        }
    }
}

@Composable
fun AchievementUnlockedDialog(
    achievement: Achievement,
    onDismiss: () -> Unit
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        icon = {
            Box(
                modifier = Modifier
                    .size(64.dp)
                    .clip(CircleShape)
                    .background(MaterialTheme.colorScheme.primary),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    Icons.Default.EmojiEvents,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onPrimary,
                    modifier = Modifier.size(32.dp)
                )
            }
        },
        title = {
            Text(
                text = "Achievement Unlocked!",
                textAlign = TextAlign.Center
            )
        },
        text = {
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(
                    text = achievement.name,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    textAlign = TextAlign.Center
                )
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = achievement.description,
                    style = MaterialTheme.typography.bodyMedium,
                    textAlign = TextAlign.Center,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    text = "+${achievement.xpReward} XP",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.primary
                )
            }
        },
        confirmButton = {
            Button(onClick = onDismiss) {
                Text("Awesome!")
            }
        }
    )
}

File: iosApp/GoSteady/Sources/Screens/Gamification/GamificationScreen.swift
import SwiftUI

struct GamificationScreen: View {
    @State private var viewModel = GamificationViewModel()
    @State private var showAchievementAlert = false
    @State private var alertAchievement: Achievement?
    @Environment(\.modelContext) private var modelContext

    var body: some View {
        ScrollView {
            LazyVStack(spacing: 16) {
                if let gamification = viewModel.gamification {
                    LevelCard(gamification: gamification)
                }

                if let rings = viewModel.todayRings {
                    DailyRingsCard(rings: rings)
                }

                if let gamification = viewModel.gamification {
                    StreakCard(gamification: gamification)
                }

                WeeklyProgressCard(weeklyRings: viewModel.weeklyRings)

                if !viewModel.recentXp.isEmpty {
                    RecentXpCard(transactions: viewModel.recentXp)
                }
            }
            .padding(16)
        }
        .navigationTitle("Progress")
        .onAppear {
            viewModel.configure(gamificationRepository: GamificationRepository(context: modelContext))
        }
        .refreshable { viewModel.loadData() }
        .onChange(of: viewModel.newAchievements.count) {
            if let first = viewModel.newAchievements.first {
                alertAchievement = first
                showAchievementAlert = true
            }
        }
        .alert("Achievement Unlocked!", isPresented: $showAchievementAlert) {
            Button("Awesome!") {
                if let a = alertAchievement {
                    viewModel.dismissAchievementNotification(a.id)
                }
            }
        } message: {
            if let a = alertAchievement {
                Text("\(a.name)\n\(a.achievementDescription)\n+\(a.xpReward) XP")
            }
        }
    }
}

// MARK: - Level Card

private struct LevelCard: View {
    let gamification: UserGamification

    var body: some View {
        VStack(spacing: 12) {
            ZStack {
                Circle()
                    .fill(AppTheme.primary)
                    .frame(width: 80, height: 80)
                Text("\(gamification.currentLevel)")
                    .font(.largeTitle.bold())
                    .foregroundStyle(.white)
            }

            Text("Level \(gamification.currentLevel)")
                .font(.title3.bold())

            Text("\(gamification.totalXp) XP")
                .font(.subheadline)
                .foregroundStyle(.secondary)

            VStack(spacing: 4) {
                HStack {
                    Text("Next Level")
                        .font(.caption2)
                    Spacer()
                    Text("\(gamification.xpToNextLevel) XP to go")
                        .font(.caption2)
                }
                ProgressView(value: gamification.levelProgress)
                    .tint(AppTheme.primary)
            }
        }
        .padding(20)
        .background(AppTheme.primaryContainer, in: RoundedRectangle(cornerRadius: 16))
    }
}

// MARK: - Daily Rings Card (Apple Watch-style concentric rings)

private struct DailyRingsCard: View {
    let rings: DailyRings

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Today's Goals")
                .font(.headline)

            HStack(spacing: 0) {
                Spacer()
                ConcentricRingsView(
                    moveProgress: CGFloat(rings.moveProgress),
                    exerciseProgress: CGFloat(rings.exerciseProgress),
                    distanceProgress: CGFloat(rings.distanceProgress)
                )
                .frame(width: 140, height: 140)
                Spacer()
            }

            HStack {
                RingLegend(color: Color(hex: "FF2D55"), label: "Move",
                           value: "\(rings.moveMinutes)/\(rings.moveGoal) min")
                Spacer()
                RingLegend(color: Color(hex: "30D158"), label: "Exercise",
                           value: "\(rings.exerciseMinutes)/\(rings.exerciseGoal) min")
                Spacer()
                RingLegend(color: Color(hex: "007AFF"), label: "Distance",
                           value: String(format: "%.1f/%.1f km",
                                         rings.distanceMeters / 1000,
                                         rings.distanceGoal / 1000))
            }
        }
        .padding(16)
        .background(AppTheme.surfaceContainer, in: RoundedRectangle(cornerRadius: 16))
    }
}

private struct RingLegend: View {
    let color: Color
    let label: String
    let value: String

    var body: some View {
        VStack(spacing: 2) {
            Circle().fill(color).frame(width: 8, height: 8)
            Text(label).font(.caption2.bold())
            Text(value).font(.caption2).foregroundStyle(.secondary)
        }
    }
}

// MARK: - Concentric Activity Rings

struct ConcentricRingsView: View {
    let moveProgress: CGFloat
    let exerciseProgress: CGFloat
    let distanceProgress: CGFloat

    @State private var animatedMove: CGFloat = 0
    @State private var animatedExercise: CGFloat = 0
    @State private var animatedDistance: CGFloat = 0

    var body: some View {
        ZStack {
            // Outer ring - Move (red)
            ActivityRingShape(progress: animatedMove, lineWidth: 14)
                .foregroundStyle(Color(hex: "FF2D55"))
                .padding(0)

            ActivityRingBackground(lineWidth: 14)
                .foregroundStyle(Color(hex: "FF2D55").opacity(0.2))
                .padding(0)

            // Middle ring - Exercise (green)
            ActivityRingShape(progress: animatedExercise, lineWidth: 14)
                .foregroundStyle(Color(hex: "30D158"))
                .padding(18)

            ActivityRingBackground(lineWidth: 14)
                .foregroundStyle(Color(hex: "30D158").opacity(0.2))
                .padding(18)

            // Inner ring - Distance (blue)
            ActivityRingShape(progress: animatedDistance, lineWidth: 14)
                .foregroundStyle(Color(hex: "007AFF"))
                .padding(36)

            ActivityRingBackground(lineWidth: 14)
                .foregroundStyle(Color(hex: "007AFF").opacity(0.2))
                .padding(36)
        }
        .onAppear {
            withAnimation(.easeOut(duration: 1.0)) {
                animatedMove = min(moveProgress, 1)
                animatedExercise = min(exerciseProgress, 1)
                animatedDistance = min(distanceProgress, 1)
            }
        }
    }
}

private struct ActivityRingShape: Shape {
    var progress: CGFloat
    var lineWidth: CGFloat

    var animatableData: CGFloat {
        get { progress }
        set { progress = newValue }
    }

    func path(in rect: CGRect) -> Path {
        var path = Path()
        let center = CGPoint(x: rect.midX, y: rect.midY)
        let radius = min(rect.width, rect.height) / 2 - lineWidth / 2
        path.addArc(center: center, radius: radius,
                    startAngle: .degrees(-90),
                    endAngle: .degrees(-90 + 360 * Double(progress)),
                    clockwise: false)
        return path.strokedPath(.init(lineWidth: lineWidth, lineCap: .round))
    }
}

private struct ActivityRingBackground: Shape {
    var lineWidth: CGFloat

    func path(in rect: CGRect) -> Path {
        var path = Path()
        let center = CGPoint(x: rect.midX, y: rect.midY)
        let radius = min(rect.width, rect.height) / 2 - lineWidth / 2
        path.addArc(center: center, radius: radius,
                    startAngle: .degrees(0), endAngle: .degrees(360),
                    clockwise: false)
        return path.strokedPath(.init(lineWidth: lineWidth, lineCap: .round))
    }
}

// MARK: - Streak Card

private struct StreakCard: View {
    let gamification: UserGamification

    var body: some View {
        HStack(spacing: 16) {
            ZStack {
                Circle()
                    .fill(Color.orange.opacity(0.2))
                    .frame(width: 56, height: 56)
                Text("\u{1F525}")
                    .font(.system(size: 28))
            }

            VStack(alignment: .leading, spacing: 2) {
                Text("\(gamification.currentStreak) Day Streak")
                    .font(.headline)
                Text(gamification.currentStreak > 0 ? "Keep it going!" : "Start your streak today!")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            Spacer()

            VStack(alignment: .trailing, spacing: 2) {
                Text("Best").font(.caption).foregroundStyle(.secondary)
                Text("\(gamification.longestStreak)")
                    .font(.title3.bold())
            }
        }
        .padding(16)
        .background(
            gamification.currentStreak > 0
                ? Color.orange.opacity(0.1)
                : AppTheme.surfaceContainer,
            in: RoundedRectangle(cornerRadius: 16)
        )
    }
}

// MARK: - Weekly Progress

private struct WeeklyProgressCard: View {
    let weeklyRings: [DailyRings]

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("This Week").font(.headline)

            HStack {
                ForEach(sortedRings, id: \.date) { ring in
                    let isToday = Calendar.current.isDateInToday(ring.date)
                    VStack(spacing: 4) {
                        Text(dayLetter(for: ring.date))
                            .font(.caption2)
                            .foregroundStyle(isToday ? AppTheme.primary : .secondary)

                        ZStack {
                            Circle()
                                .fill(circleColor(ring: ring, isToday: isToday))
                                .frame(width: 32, height: 32)
                            if ring.allRingsClosed {
                                Image(systemName: "checkmark")
                                    .font(.caption2.bold())
                                    .foregroundStyle(.white)
                            }
                        }
                    }
                    if ring.date != sortedRings.last?.date {
                        Spacer()
                    }
                }
            }
        }
        .padding(16)
        .background(AppTheme.surfaceContainer, in: RoundedRectangle(cornerRadius: 16))
    }

    private var sortedRings: [DailyRings] {
        weeklyRings.sorted { $0.date < $1.date }
    }

    private func dayLetter(for date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEEE"
        return formatter.string(from: date)
    }

    private func circleColor(ring: DailyRings, isToday: Bool) -> Color {
        if ring.allRingsClosed { return Color(hex: "30D158") }
        if ring.exerciseMinutes > 0 { return Color.orange.opacity(0.5) }
        if isToday { return AppTheme.primaryContainer }
        return AppTheme.surfaceContainer
    }
}

// MARK: - Recent XP

private struct RecentXpCard: View {
    let transactions: [XpTransaction]

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Recent XP").font(.headline)

            ForEach(Array(transactions.prefix(5))) { tx in
                HStack {
                    Text(tx.reason.displayName)
                        .font(.subheadline)
                    Spacer()
                    Text("+\(tx.amount) XP")
                        .font(.subheadline.bold())
                        .foregroundStyle(AppTheme.primary)
                }
                .padding(.vertical, 2)
            }
        }
        .padding(16)
        .background(AppTheme.surfaceContainer, in: RoundedRectangle(cornerRadius: 16))
    }
}

#Preview {
    NavigationStack {
        GamificationScreen()
    }
}
